### ISSUE TITLE
Memory leak: tf1 trained saved_model in tf2 for prediction

### ISSUE THREAD

**Issue by purvang3** [tensorflow/tensorflow#57601]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.6

### Custom Code

Yes

### OS Platform and Distribution

Ubuntu 18.04

### Python version

3.6.9

### GCC/Compiler version

tensorflow  docker image

### CUDA/cuDNN version

tensorflow docker image

### GPU model and memory

tensorflow docker image

### Current Behaviour?

```shell
A bug happened!

I have previously trained ssd_inception_v2 model in tensorflow 1.14.
It has frozen_inference graph and saved_model dir with protobuf files and variables.
I am running tensorflow 2.6.0. loading tf 1.14 trained saved_model into tf 2.6
is done without problem and it runs smoothly. But over the period of time, 
cpu memory keeps increasing and after some time, prediction scrip crashes 
because of memory full. I have tried to load "frozen graph.pb" instead of 
saved_model.pb and problem still exists. Any help would be appreciated.
Using "htop" command, MEM% column keep increasing over the
time with following script running.
```

### Standalone code to reproduce the issue

Any tensorflow 1 trained model with saved_model dir after training.
sample: wget http://download.tensorflow.org/models/object_detection/ssd_inception_v2_coco_2018_01_28.tar.gz

use saved_model dir.

Docker image: docker pull tensorflow/tensorflow:2.6.0-gpu-jupyter
```shell
import numpy as np

import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()
tf.config.set_soft_device_placement(True)
output_tensor_dict_global = None

model_path = "saved_model_path"

gpus = tf.config.experimental.list_physical_devices('GPU')

def get_output_tensor_dict():
    global output_tensor_dict_global
    if output_tensor_dict_global:
        return output_tensor_dict_global

    ops = tf.get_default_graph().get_operations()
    all_tensor_names = {output.name for op in ops for output in op.outputs}
    tensor_dict = {}
    for key in ['num_detections', 'detection_boxes', 'detection_scores', 'detection_classes', 'detection_masks']:
        tensor_name = key + ':0'
        if tensor_name in all_tensor_names:
            tensor_dict[key] = tf.get_default_graph().get_tensor_by_name(
                tensor_name)
    output_tensor_dict_global = tensor_dict
    return output_tensor_dict_global


if gpus:
    try:
        for gpu in gpus:
            tf.config.experimental.set_memory_growth(gpu, True)
    except RuntimeError as e:
        print(e)

with tf.Graph().as_default() as g:
    with tf.Session() as sess:
        detection_graph = tf.saved_model.load(sess, ["serve"], model_path)

        while True:
            img_np = np.random.randn(720,1280,3)
            tensor_dict = get_output_tensor_dict()
            image_tensor = tf.get_default_graph().get_tensor_by_name('image_tensor:0')
            output_dict = sess.run(tensor_dict, feed_dict={image_tensor: np.expand_dims(img_np, axis=0)})
            print(output_dict.keys())
          
```
I have tested same code with tensorflow 2.9.0 and problem still exists.
```


### Relevant log output

```shell
Memory consumption should be constant.
```
</details>

**Comment by sushreebarsa** [comment]
@purvang3 
In order to expedite the trouble-shooting process, please provide a complete code snippet to reproduce the issue reported here. Thank you!

**Comment by purvang3** [comment]
```
import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()

import numpy as np
tf.config.set_soft_device_placement(True)

model_path = ""

gpus = tf.config.experimental.list_physical_devices('GPU')
if gpus:
    try:
         for gpu in gpus:
              tf.config.experimental.set_memory_growth(gpu, True)
    except RuntimeError as e:
         print(e)

with tf.Graph().as_default() as g:
        with tf.Session() as sess:
            detection_graph = tf.saved_model.load(sess, ["serve"], model_path)
            while True:
                img_np = np.random.randn(720,1280,3)
                img_np = tf.cast([img_np], dtype=tf.uint8)
                output_dict = detection_graph.signatures[tf.saved_model.DEFAULT_SERVING_SIGNATURE_DEF_KEY](img_np)
                output_dict = run_inference_for_single_image(sess, image)
                output_dict[c.NUM_DETECTIONS_KEY] = int(output_dict["num_detections"].numpy())
                output_dict[c.BOXES_KEY] = output_dict["detection_boxes"][0].numpy()
                output_dict[c.KLASSES_KEY] = output_dict["detection_classes"][0].numpy()
                output_dict[c.SCORES_KEY] = output_dict["detection_scores"][0].numpy()
```

@sushreebarsa The attached code snippet is the entire code that I run to analyze memory. Only thing user need to do to download the model from   http://download.tensorflow.org/models/object_detection/ssd_inception_v2_coco_2018_01_28.tar.gz and add the model path in code.

Thank you

**Comment by sushreebarsa** [comment]
@purvang3 Thank you for the update! 
This issue seems to be more related to models. Could you please post this issue on [models](https://github.com/tensorflow/models/issues?q=is%3Aissue+is%3Aopen) repo to get the right help there?
Thank you!

**Comment by purvang3** [comment]
@sushreebarsa yes, it's a same issue. Could you please explain how does it related to models and not tensorflow? just curious.

**Comment by tiruk007** [comment]
@purvang3 
I reproduced the code shared but facing different error. Could you please go through the given gist and confirm it. Kindly find gist [here](https://colab.research.google.com/gist/tiruk007/1668be10f5a741979593008ec3bab8b8/57601.ipynb)
Thank you!

**Comment by purvang3** [comment]
@tiruk007 I have tried the same code in google colab as well and it's working without any error. it comes with tensorflow version 2.8. when you install new version, sometimes it's better to restart runtime. Here is the gist link (https://colab.research.google.com/drive/1XiYFxsBq4nN5eN30AT27Y4oSCuHQpa1p?usp=sharing). please check it out. Thank you

**Comment by tiruk007** [comment]
@gadagashwini  the issue Was able to reproduce the issue on Colab using TF v2.9 Please find the gist [here](https://colab.research.google.com/gist/tiruk007/4c969c5fa4cced94286a6e3a38bdc70f/57601.ipynb) for reference .
Thank you!

**Comment by gadagashwini** [comment]
Hi @purvang3, Did you followed the instructions mention https://www.tensorflow.org/guide/migrate/saved_model#tensorflow_1_load_a_savedmodel_with_tfsaved_modelload. Thank you!

**Comment by purvang3** [comment]
@gadagashwini . Thank you for your reply. I have followed the same guide to load the saved_model and do inference and memory leak problem still exists.

**Comment by sachinprasadhs** [comment]
Is it possible for you to try saving the model with latest Tensorflow version and load it using `model.save` and load the model using `tf.keras.models.load_model` and let us know if you still observe the memory leak. Thanks!

**Comment by sachinprasadhs** [comment]
> Saving it using the newest version of Tensorflow 1? It only occurs when saving a model in Tensorflow 1 and I tried with Tensorflow 1.15 which should be the newest version.

Try saving the model in Tensorflow version 2.5 and above, for migrating your 1.x code to 2.x, you can follow https://www.tensorflow.org/guide/migrate

**Comment by rrajkumar1990** [comment]
Hi Team ,  I am also facing a similar issue .  I have a tensorflow v1  trained model with Retinanet architecture .  The model  training and inference code have been  ported to tensorflow v2 through the compat.v1 apis.

Issue : 
 I run my model inference inside a docker container with memory 1.5gb.
When I load the model in memory , and  I could see that the docker memory is getting occupied (around 1gb). 
With the cached model , during  inference for a single image  I see that there is additional increase in the memory (container memory is now at  1.2gb) .  The expectation is , the  model should  release the memory used for inference .And the memory consumption of the container shud be only ~1GB (model cache) after the inference
But the memory doesn’t get released  and when  the cached model is used for multiple  inference calls sequentially , the memory consumption reaches the max memory and the container gets SIGKILL error .

Versions: 
I see the same behaviour for Tensorflow  2.8.0 and 2.9.1 versions.

**Comment by muxamilian** [comment]
> Is it possible for you to try saving the model with latest Tensorflow version and load it using `model.save` and load the model using `tf.keras.models.load_model` and let us know if you still observe the memory leak. Thanks!

I tried saving the tf1 model using tf2.10 but the memory leak persists. 

Then I tried loading the saved model in tf2.10 again and saving it one more time with tf2.10 but then I get another problem: The newly saved model only is a couple of KB in size. Apparently all neural network weights are lost. 
```python
model = tf.saved_model.load(base_path+"/tf2_saved_model_test")
tf.saved_model.save(model, base_path+"/tf2_saved_model_test2")
```

**Comment by sachinprasadhs** [comment]
@muxamilian , Did you try saving the model using model.save and load it using tf.keras.models.load_model.

You can also try saving model in .keras format using nightly version.
Below is the save and load format.

```
model = tf.keras.Sequential([
            tf.keras.layers.Dense(5, input_shape=(3,)),
            tf.keras.layers.Softmax()])
        model.save("model.keras")
        loaded_model = tf.keras.models.load_model("model.keras")

```

**Comment by muxamilian** [comment]
I assume you're wondering why I (and others) still use tf1. The reason is that for the Google Coral, all the official state-of-the-art object detection models (like [MobileDet](https://colab.research.google.com/github/google-coral/tutorials/blob/master/retrain_ssdlite_mobiledet_qat_tf1.ipynb) or [SpaghettiNet](https://discuss.tensorflow.org/t/training-a-spaghettinet-model/8648/9) are still trained using tf1. 

Thus I train using tf1 to make the models work on the Coral. However, I also want to run them on the GPU and that's why I convert them to tf2 later. And then the memory leak occurs. Hope this clarified it.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
